File: src/keelson/campaign/scheduler.py
"""Simple campaign scheduler — foreground asyncio loop with cron-like timing."""

from __future__ import annotations

import asyncio
import re
from collections.abc import Callable
from pathlib import Path

from keelson.adapters.base import BaseAdapter
from keelson.campaign.runner import run_campaign
from keelson.core.models import CampaignConfig, CampaignResult, Target


def parse_interval(interval: str) -> int:
    """Parse a human-readable interval string into seconds.

    Supports: "30s", "5m", "1h", "2h30m", "1d"
    """
    total = 0
    pattern = re.compile(r"(\d+)([smhd])")
    matches = pattern.findall(interval.lower())
    if not matches:
        raise ValueError(f"Invalid interval format: {interval}. Use e.g. '30s', '5m', '1h', '1d'")
    for value, unit in matches:
        n = int(value)
        if unit == "s":
            total += n
        elif unit == "m":
            total += n * 60
        elif unit == "h":
            total += n * 3600
        elif unit == "d":
            total += n * 86400
    return total


async def run_scheduled(
    target: Target,
    adapter: BaseAdapter,
    config: CampaignConfig,
    interval_seconds: int,
    max_runs: int | None = None,
    attacks_dir: Path | None = None,
    on_campaign: Callable[[CampaignResult, int], None] | None = None,
) -> list[CampaignResult]:
    """Run campaigns on a schedule.

    Args:
        target: Target to scan.
        adapter: Target adapter.
        config: Campaign configuration.
        interval_seconds: Seconds between campaign runs.
        max_runs: Maximum number of runs (None = run forever).
        attacks_dir: Override probes directory.
        on_campaign: Callback(result, run_number) after each campaign.

    Returns:
        List of all campaign results.
    """
    results: list[CampaignResult] = []
    run_number = 0

    while max_runs is None or run_number < max_runs:
        run_number += 1
        result = await run_campaign(target, adapter, config, attacks_dir=attacks_dir)
        results.append(result)

        if on_campaign:
            on_campaign(result, run_number)

        if max_runs is not None and run_number >= max_runs:
            break

        await asyncio.sleep(interval_seconds)

    return results

File: src/campaign/scheduler.ts
/**
 * Campaign scheduler -- foreground async loop with configurable timing.
 *
 * Ports _legacy/src/campaign/scheduler.py to TypeScript.
 */

import { runCampaign } from './runner.js';
import type { RunCampaignOptions } from './runner.js';
import type { Adapter, CampaignConfig, CampaignResult } from '../types/index.js';

// ─── Interval Parsing ───────────────────────────────────

const UNIT_MS: Record<string, number> = {
  s: 1_000,
  m: 60_000,
  h: 3_600_000,
  d: 86_400_000,
};

const INTERVAL_PATTERN = /(\d+)([smhd])/g;

/**
 * Parse a human-readable interval string into milliseconds.
 *
 * Supports: "30s", "5m", "1h", "2d", and compound forms like "2h30m".
 */
export function parseInterval(interval: string): number {
  const matches = [...interval.toLowerCase().matchAll(INTERVAL_PATTERN)];

  if (matches.length === 0) {
    throw new Error(
      `Invalid interval format: "${interval}". Use e.g. "30s", "5m", "1h", "2h30m", "1d".`,
    );
  }

  let totalMs = 0;
  for (const [, value, unit] of matches) {
    totalMs += Number(value) * UNIT_MS[unit];
  }

  return totalMs;
}

// ─── Scheduled Runner ───────────────────────────────────

export type OnCampaignCallback = (result: CampaignResult, runNumber: number) => void;

export interface ScheduledRunOptions extends RunCampaignOptions {
  /** Milliseconds between campaign runs. */
  intervalMs: number;
  /** Maximum number of runs. Omit or set undefined to run until aborted. */
  maxRuns?: number;
  /** AbortSignal for graceful shutdown. */
  signal?: AbortSignal;
  /** Called after each campaign completes. */
  onCampaign?: OnCampaignCallback;
}

/**
 * Run campaigns on a schedule.
 *
 * Executes `runCampaign` in a loop, sleeping `intervalMs` between runs.
 * Stops when `maxRuns` is reached or the `signal` is aborted.
 *
 * @returns All campaign results collected during the run.
 */
export async function runScheduled(
  target: string,
  adapter: Adapter,
  config: CampaignConfig,
  options: ScheduledRunOptions,
): Promise<CampaignResult[]> {
  const { intervalMs, maxRuns, signal, onCampaign, ...campaignOptions } = options;
  const results: CampaignResult[] = [];
  let runNumber = 0;

  while (maxRuns === undefined || runNumber < maxRuns) {
    if (signal?.aborted) break;

    runNumber++;
    const result = await runCampaign(target, adapter, config, campaignOptions);
    results.push(result);

    onCampaign?.(result, runNumber);

    if (maxRuns !== undefined && runNumber >= maxRuns) break;
    if (signal?.aborted) break;

    await delay(intervalMs, signal);
  }

  return results;
}

// ─── Helpers ────────────────────────────────────────────

/**
 * Promise-based delay that resolves after `ms` milliseconds or when the
 * AbortSignal fires, whichever comes first.
 */
function delay(ms: number, signal?: AbortSignal): Promise<void> {
  if (signal?.aborted) return Promise.resolve();

  return new Promise<void>((resolve) => {
    const timer = setTimeout(() => {
      cleanup();
      resolve();
    }, ms);

    function onAbort(): void {
      clearTimeout(timer);
      cleanup();
      resolve();
    }

    function cleanup(): void {
      signal?.removeEventListener('abort', onAbort);
    }

    signal?.addEventListener('abort', onAbort, { once: true });
  });
}
